src/processor/operator/persistent/reader/csv/serial_csv_reader.cpp

```diff
@@ -17,21 +16,25 @@ SerialCSVReader::SerialCSVReader(const std::string& filePath, CSVOption option,
 
 std::vector<std::pair<std::string, LogicalType>> SerialCSVReader::sniffCSV() {
     readBOM();
-
-    if (option.hasHeader) {
-        SniffCSVNameAndTypeDriver driver{context};
-        parseCSV(driver);
-        return std::move(driver.columns);
-    } else {
-        SniffCSVColumnCountDriver driver;
-        parseCSV(driver);
-        std::vector<std::pair<std::string, LogicalType>> columns;
-        columns.reserve(driver.numColumns);
-        for (uint64_t i = 0; i < driver.numColumns; i++) {
-            columns.emplace_back(stringFormat("column{}", i), LogicalTypeID::STRING);
+    SniffCSVNameAndTypeDriver driver{context, option, this};
+    parseCSV(driver);
+    // finalize the columns; rename duplicate names
+    std::map<std::string, int32_t> names;
+    for (auto& i : driver.columns) {
+        // Suppose name "col" already exists
+        // Let N be the number of times it exists
+        // rename to "col" + "_{N}"
+        // ideally "col_{N}" shouldn't exist, but if it already exists M times (due to user
+        // declaration), rename to "col_{N}" + "_{M}" repeat until no match exists
+        while (names.contains(i.first)) {
+            names[i.first]++;
+            i.first += "_" + std::to_string(names[i.first]);
         }
-        return columns;
+        names[i.first];
+        // purge null types
+        i.second = LogicalTypeUtils::purgeAny(i.second, LogicalType::STRING());
     }
+    return std::move(driver.columns);
```

Comment:
I'm pretty sure this std::move is unnecessary. U don't get any performant benefit from this move, in fact it may even mislead compiler in some cases and miss potential optimization.

Reply:
You're right

Reply:
Nevermind, std::move is necessary because you can't copy logical types.